lib/dynarray.c

```diff
@@ -0,0 +1,184 @@
+ *      Pittsburgh, PA  15213
+ *      (412) 268-7393, fax: (412) 268-7395
+ *      innovation@andrew.cmu.edu
+ *
+ * 4. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by Computing Services
+ *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
+ *
+ * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
+ * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
+ * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <assert.h>
+#include <memory.h>
+
+#include "dynarray.h"
+#include "util.h"
+#include "xmalloc.h"
+
+EXPORTED void dynarray_init(struct dynarray *da, size_t membsize)
+{
+    assert(membsize);
+    memset(da, 0, sizeof(struct dynarray));
+    da->membsize = membsize;
+}
+
+EXPORTED void dynarray_fini(struct dynarray *da)
+{
+    free(da->data);
+    memset(da, 0, sizeof(struct dynarray));
+}
+
+EXPORTED struct dynarray *dynarray_new(size_t membsize)
+{
+    struct dynarray *da = xmalloc(sizeof(struct dynarray));
+    dynarray_init(da, membsize);
+    return da;
+}
+
+EXPORTED void dynarray_free(struct dynarray **dap)
+{
+    if (!dap || *dap == NULL) return;
+    free((*dap)->data);
+    free(*dap);
+    *dap = NULL;
+}
+
+/*
+ * Ensure the index @newalloc exists in the array, if necessary expanding the
+ * array, and if necessary NULL-filling all the intervening elements.
+ * Note that we always ensure an empty slot past the last reported
+ * index, so that we can pass data[] to execve() or other routines that
+ * assume a NULL terminator.
+ */
+#define QUANTUM     16
+static void ensure_alloc(struct dynarray *da, int newalloc)
+{
+    assert(newalloc >= 0);
+    if (newalloc)
+        newalloc++;
+    if (newalloc <= da->alloc)
+        return;
+    newalloc = ((newalloc + QUANTUM-1) / QUANTUM) * QUANTUM;
+    da->data = xrealloc(da->data, da->membsize * newalloc);
+    memset(da->data + da->alloc * da->membsize, 0, da->membsize * (newalloc-da->alloc));
+    da->alloc = newalloc;
+}
+
+
+static inline int adjust_index_ro(const struct dynarray *da, int idx)
+{
+    if (idx >= da->count)
+        return -1;
+    else if (idx < 0)
+        idx += da->count;
+    return idx;
+}
+
+static inline int adjust_index_rw(struct dynarray *da, int idx, int len)
+{
+    if (idx >= da->count) {
+        ensure_alloc(da, idx+len);
+    } else if (idx < 0) {
+        idx += da->count;
+        if (idx >= 0 && len)
+            ensure_alloc(da, da->count+len);
+    } else if (len) {
+        ensure_alloc(da, da->count+len);
+    }
+    return idx;
+}
+
+EXPORTED int dynarray_size(struct dynarray *da)
```

Comment:
Can the function `dynarray_size()` be converted to macro in dynarray.h?  This saves a function call any time the size is requested.

Likewise for `arrayu64_size()`, `ptrarray_size()`, `strarray_size()`, `bufarray_size()`.